src/lightgbm/src/main/scala/LightGBMUtils.scala

```diff
@@ -37,6 +45,50 @@ object LightGBMUtils {
     featurizer.fit(dataset)
   }
 
+  /**
+    * Opens a socket communications channel on the driver, starts a thread that
+    * waits for the host:port from the executors, and then sends back the
+    * information to the executors.
+    * @param numExecutorCores The total number of executors * cores to wait for.
+    * @return The address and port of the driver socket.
+    */
+  def createDriverNodesThread(numExecutorCores: Int, df: DataFrame, log: Logger): (String, Int) = {
+    val numPartitions = df.rdd.getNumPartitions
+    val numWorkers = math.min(numExecutorCores, numPartitions)
+    // Start a thread and open port to listen on
+    implicit val context = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
+    val driverServerSocket = new ServerSocket(0)
+    val f = Future {
+      val hostAndPorts = ListBuffer[(Socket, String)]()
+      log.info(s"driver expecting $numWorkers connections...")
+      while (hostAndPorts.size < numWorkers) {
+        log.info("driver accepting a new connection...")
+        val driverSocket = driverServerSocket.accept()
+        val reader = new BufferedReader(new InputStreamReader(driverSocket.getInputStream))
+        val comm = reader.readLine()
+        log.info(s"driver received socket from worker: $comm")
+        val socketAndComm = (driverSocket, comm)
+        hostAndPorts += socketAndComm
+      }
+      // Concatenate with commas, eg: host1:port1,host2:port2, ... etc
+      val allConnections = hostAndPorts.map(_._2).mkString(",")
+      log.info(s"driver writing back to all connections: $allConnections")
+      // Send data back to all threads on executors
+      hostAndPorts.foreach(hostAndPort => {
+        val writer = new BufferedWriter(new OutputStreamWriter(hostAndPort._1.getOutputStream))
+        writer.write(allConnections + "\n")
+        writer.flush()
+      })
+      log.info("driver closing all sockets and server socket")
+      hostAndPorts.foreach(_._1.close())
+      driverServerSocket.close()
+    }
+    val host = getDriverHost(df)
+    val port = driverServerSocket.getLocalPort
+    log.info(s"driver waiting for connections on host: ${host} and port: $port")
```

Comment:
Await on this future to ensure it finishes

Reply:
done